pulp_container/app/registry_api.py

```diff
@@ -473,6 +477,135 @@ def get_queryset(self, *args, **kwargs):
         return (public_repositories | accessible_repositories).distinct()
 
 
+class FlatpakIndexDynamicView(APIView):
+    """
+    Handles requests to the /index/dynamic endpoint
+    """
+
+    authentication_classes = []
+    permission_classes = []
+
+    def recurse_through_manifest_lists(self, tag, manifest, oss, architectures, manifests):
+        mlms = manifest.listed_manifests.through.objects.filter(image_manifest__pk=manifest.pk)
+        if oss:
+            mlms.filter(os__in=oss)
+        if architectures:
+            mlms.filter(architecture__in=architectures)
+        for mlm in mlms:
+            self.recurse_through_manifest_lists(
+                tag, mlm.manifest_list, oss, architectures, manifests
+            )
+        if manifest.media_type in (models.MEDIA_TYPE.MANIFEST_V2, models.MEDIA_TYPE.MANIFEST_OCI):
+            manifests.setdefault(manifest, set()).add(tag)
+
+    def get(self, request):
+        req_repositories = None
+        req_tags = None
+        req_oss = None
+        req_architectures = None
+        req_label_exists = set()
+        req_label_values = {}
+        for key, values in request.query_params.lists():
+            if key == "repository":
+                req_repositories = values
+            elif key == "tag":
+                req_tags = values
+            elif key == "os":
+                req_oss = values
+            elif key == "architecture":
+                req_architectures = values
+            elif key.startswith("label:"):
+                if key.endswith(":exists"):
+                    if any(v != "1" for v in values):
+                        raise ParseError()
+                    label = key[len("label:") : len(key) - len(":exists")]
+                    req_label_exists.add(label)
+                else:
+                    label = key[len("label:") :]
+                    req_label_values[label] = values
+            else:
+                # In particularly, this covers any annotation:... parameters, which this
+                # implementation does not support:
+                raise ParseError()
+
+        if "org.flatpak.ref" not in req_label_exists:
+            raise ParseError()
+
+        distributions = models.ContainerDistribution.objects.filter(private=False).only("base_path")
+
+        if req_repositories:
+            distributions = distributions.filter(base_path__in=req_repositories)
+
+        results = []
+        for distribution in distributions:
+            images = []
+            if distribution.repository:
+                repository_version = distribution.repository.latest_version()
+            else:
+                repository_version = distribution.repository_version
+            tags = models.Tag.objects.filter(pk__in=repository_version.content)
```

Comment:
```suggestion
            tags = models.Tag.objects.select_related("tagged_manifest").filter(pk__in=repository_version.content)
```

Reply:
that's functionally equivalent and "just" a performance improvement?

Reply:
This should be a performance improvement.

We hit the database after doing `for tag in tags` first. Then, we hit the database every iteration when we try to work with the tagged manifests. By selecting the related field, we load the referenced objects ahead of time.

Reply:
done in c6605fe350000eaa3c0de35d05f73af25453405d